46.md

```diff
@@ -1,98 +1,162 @@
+
+All pubkeys specified in this NIP are in hex format.
 
 ## Signer Discovery
 
-The client always starts by generating a random key which is used to communicate with the signer, then it one of the methods below is used to allow the client to know what is the signer public key for the session and which relays to use.
+To initiate a connection between a client and a remote signer there are a few different options.
 
-### Started by the signer (nsecBunker)
+### Direct connection initiated by remote signer
 
-The remote signer generates a connection token in the form
+This is most common in a situation where you have your own nsecbunker or other type of remote signer and want to connect through a client that supports remote signing.
+
+The remote signer would provide a connection token in the form:
 
 ```
-bunker://<hex-pubkey>?relay=wss://...&relay=wss://...&secret=<optional-secret>
+bunker://<remote-pubkey>?relay=<wss://relay-to-connect-on>&relay=<wss://another-relay-to-connect-on>&secret=<optional-secret-value>
 ```
 
-The user copies that token and pastes it in the client UI somehow. Then the client can send events of kind `24133` to the specified relays and wait for responses from the remote signer.
+This token is pasted into the client by the user and the client then uses the details to connect to the remote signer via the specified relay(s).
 
-### Started by the client
+### Direct connection initiated by the client
 
-The client generates a QR code in the following form (URL-encoded):
+In this case, basically the opposite direction of the first case, the client provides a connection token (or encodes the token in a QR code) and the signer initiates a connection to the client via the specified relay(s).
 
 ```
-nostrconnect://<client-key-hex>?relay=wss://...&metadata={"name":"...", "url": "...", "description": "..."}
+nostrconnect://<local-keypair-pubkey>?relay=<wss://relay-to-connect-on>&metadata=<json metadata in the form: {"name":"...", "url": "...", "description": "..."}>
 ```
 
-The signer scans the QR code and sends a `connect` message to the client in the specified relays.
+### Remote signer discovery via NIP-89
+
+In this last case, most often used to fascilitate an OAuth-like signin flow, the client first looks for remote signers that have announced themselves via NIP-89 application handler events.
+
+First the client will query for `kind: 31989` events that have a `k` tag of `24133`.
+
+These are generally shown to a user, and once the user selects which remote signer to use and provides the remote pubkey they want to use (via npub, pubkey, or nip-05 value), the client can initiate a connection. Note that it's on the user to select the remote signer that is actually managing the remote key that they would like to use in this case. If the remote pubkey is managed on another remote signer, the connection will fail.
```

Comment:
> and provides the remote pubkey they want to use (via npub, pubkey, or nip-05 value)

This is a viable flow, but not I think the best UX. Best would be for a user to provide a username + domain so users don't have to remember their pubkey. Providing a different nip05 than the one you're logging in to would be very confusing. If this is a "create account" flow, the signer should return an error if the username is already taken.

But an optimistic connection request should also be supported. This would work the same as the `create_account` request, except that whether the username exists or not, the signer should return an auth_url rather than an error. If the username exists, the auth_url would be straightforward login screen. If the username doesn't exist, the signer should display a modified sign up screen with an option to add an existing private key. To make it clear, the `create_account` verb could be renamed to `connect_with_username` or something. The idea is that signers should follow postel's law and do their best not to return an error if possible.

Here's how I would write this:

```suggestion
The client MAY display the list of domains to the user so they can select the signer they prefer to use. Clients SHOULD allow users to enter a custom domain in case there is a signer that wasn't discovered.

In addition to the signer domain, the user must provide some way to identify their account, any of:

- Pubkey/npub
- NIP 05 address (even if the domain differs from the signer)
- Username (recommended)

If a NIP 05 address is provided, the client should de-reference it to find the user's pubkey and display an error if no record exists. If a pubkey is provided (either via NIP 05 address or directly), the client should then attempt to `connect` with the signer by p-tagging the provided **user pubkey**.

If a bare username is provided, the client should send a `connect_with_username` request p-tagging the **signer pubkey**. The signer should then return an `auth_url` which the client can use to redirect the user to complete the process. This auth_url should be a sign in page if the username exists in the signer's database, and a sign up page if the username doesn't exist yet. If desired, the signer MAY provide a field on the sign up form that allows users to enter an existing private key to custody. A client may exert more control over this flow if desired by checking the signer's `nostr.json` for a record containing the provided username, and showing a different form client-side before redirecting to the signer.

The username-based flow is optional, clients should be prepared for signers to omit support for it, either by having no publicly accessible domain, or by ignoring/rejecting the `connect_with_username` request.
```